subprojects/base-annotations/src/main/java/org/gradle/internal/service/scopes/ServiceScope.java

```diff
@@ -31,6 +31,6 @@
 @Inherited
 public @interface ServiceScope {
 
-    Scopes value();
+    Class<? extends Scope> value();
```

Comment:
One potential problem with using `Class` to represent the scope, rather than say an instance of type `Scope`, is that we can't easily add any more metadata to a scope and have consumers query it. It's certainly possible, but it's awkward. Querying the parent scopes of a given scope is made awkward by having to iterate over `getInterfaces()` and filtering non-scope types. It's also quite slow, at least relative to something like a `getParents()` method on `Scope`. Possibly we won't need these things, though. We don't at the moment.

Reply:
I agree that using an instance of `Scope` would be better, but unfortunately Java doesn't allow an instance of a custom type like `Scope` as an annotation parameter, hence the use of `Class`.

A slight advantage of using `Class` for categorization is that `instanceof` works across multiple levels of hierarchy, too. But then again, as you said we don't need it.

If we need to attach metadata, we could add it via annotations on the scopes themselves. Not nice, but I don't see any other middle ground between using enums vs being extensible.